### PR TITLE
TypeScript Rewrite

### PR DESCRIPTION
**Pull request by hipstersmoothie**
Mostly a simple rewrite from JS to TS.

The biggest changes are in `detector.ts` where `pkg` does some babel AST manipulation. To get the types to play nice I used `@babel/types` util function which simplified the code a little bit.

This PR also contains a pretty big stylistic change. There were a bunch of one line `if` statements and those really hinder readability. I also added newlines around all `if`, `for`, and `while` for the same reason. I'm willing to revert some of these changes but I think maintainability is helped by easier to scan code.

### CODE REVIEW

**Comment by leerob** [comment]
Wow, you're not messing around 😮 🔥 Incredible!

**Comment by hipstersmoothie** [comment]
🥳  The build passed

File: lib/common.ts
Comment (leerob) [inline lib/common.ts:46]: Looks like there's a few type `any` checks missing. Add them here or in a follow up? Up to you.

image

**Review by leerob — APPROVED** [review]
Tests pass 🥳

**Comment by leerob** [comment]
You still have it marked as WIP, so let me know when you're ready.

**Comment by robertsLando** [comment]
Wow love this :heart: BTW I would wait before merging, there are still some opened pr to sort out

**Comment by robertsLando** [comment]
Please don't merge this, hold on...

**Comment by hipstersmoothie** [comment]
Got rid of most of the `any`s. There are a few that make it really hard to type the code and would require lots of logic changes to make work. I think it's fine to hold off on those until someone is working on that code that knows what it does a little more

**Comment by hipstersmoothie** [comment]
This is ready but we can wait until @robertsLando give the 👍

**Comment by robertsLando** [comment]
I would wait for #981 but mostly #1060, otherwise this will introduce too much conflicts with them and users could give up. That's my opinion but if you prefer to merge this now feel free to do it :)

**Comment by hipstersmoothie** [comment]
I agree. I'm not in any hurry

**Comment by robertsLando** [comment]
@hipstersmoothie Failing tests?

**Comment by leerob** [comment]
Okay, all the PRs have finally been merged and we are ready to ship this thing after merge conflicts are resolved! 🚀

**Comment by robertsLando** [comment]
Yep agree, @hipstersmoothie once conflicts are fixed we can go on with this :)

**Review by robertsLando — APPROVED** [review]
LGTM 🚀

**Review by leerob — APPROVED** [review]

**Pull request merged**
